Exporting functionality added

adrosenbaum opened the pull request:
User can now export synthetic datasets from mutacc by making a query to
the db, and providing 'background' bam files. The synthetic dataset will
be returned as bam files enriched with variants from the query.

henrikstranneheim commented on mutacc/builds/build_dataset.py:
```diff
@@ -0,0 +1,166 @@
+            child = sample
+            break
+    try:
+        child
```
What about siblings? How would these be handled? @moonso Do you have any functions in genmod or other packages that already perform these identity checks?

adrosenbaum replied:
at the moment this only work for father-mother-child trios. I'm trying to come up with a way that would work for all cases. I thought we could discuss this in our next meeting.

moonso replied:
Yes I agree. This should be handled already when parsing the ped file. In that way you can store something like `sample.is_child=True`. You can look at github.com/moonso/ped_parser if you like.

henrikstranneheim commented on mutacc/builds/build_variant.py:
```diff
@@ -34,28 +34,28 @@ def find_region(self, padding):
+
                 #Depending on the direction of the break end, the region is extended either upstreams
                 #or downstream of the position of the break end.
                 if "[" in self.entry.ALT[0]:
```
Are the if and else condition defining the same region? Maybe I am missing something?

henrikstranneheim replied:
Also, this looks like a very good place to use a technique called a dispatch table. Look it up and see if you can refactor the function.

henrikstranneheim commented on mutacc/mutaccDB/query.py:
```diff
@@ -0,0 +1,88 @@
+                       "start": variant["reads_region"]["start"],
+                       "end": variant["reads_region"]["end"]}
+
+            if overlapping_region(region, final_regions):
```
Nice!

henrikstranneheim commented on mutacc/utils/bam_handler.py:
```diff
@@ -130,30 +153,36 @@ def out_file(self):
 
         return str(self.out_name)
 
-    #This method may be implemented using picard FilterSamReads or another faster command line
```
Have you tried bedtools or samtools?
1. bedtools intersect -abam file.bam -b filter.bed -v > filtered.bam
2. samtools view -U -L filter.bed > filtered.bam

adrosenbaum replied:
That might be a better option. The only problem is that we might want to filter out all the mates to the reads overlapping the regions as well.

moonso commented on mutacc/builds/build_dataset.py:
```diff
@@ -0,0 +1,166 @@
+        elif sample["sample_id"] == child["mother"]:
+            mother = sample
+
+    try:
```
Same as above

adrosenbaum commented:
Some changes in PR based on discussion on export functionality we had on Friday.
Now, the read names overlapping the region are found with pysam,
and excluded from the input 'background' fastq files with command
line tool seqkit, which I think works well, and fast, for the purpose.

The pedigree structure of the cases is parsed with moonso's ped_parser
using the Individual, and Family classes.

When using mutacc grep now, the user must provide option -m, stating
the family member to make a synthetic fastq file for. if -m 'affected',
A Synthetic fastq based on any affected sample from all of the cases,
regardless of single or trio, will be made.

I've made pretty large changes since the last commit, and I am sorry for
this very bulky PR which I understand is a mess to review. If you are interested
in discussing the changes let me know.
